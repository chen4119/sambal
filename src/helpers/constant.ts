
/*
export const WebElement = {
    SiteNavigation: "SiteNavigation",
    AdBlock: "AdBlock",
    Footer: "Footer",
    Header: "Header",
    SideBar: "SideBar",
    // Table: "Table"
}

export const Action = {
    Search: "Search",
    Review: "Review",
    Like: "Like",
    Listen: "Listen",
    View: "View",
    Watch: "Watch",
    Subscribe: "Subscribe",
    Comment: "Comment",
    Reply: "Reply",
    Share: "Share",
}*/

export const SORT_ASC = "asc";
export const SORT_DESC = "desc";

export type SortBy = {
    prop: string,
    order: "asc" | "desc"
}

export type Theme = {
    name: string,
    options: object
};

export type Collection = {
    "@id": string,
    "@type"?: string,
    src: string | string[],
    groupBy?: string[],
    sortBy?: SortBy
}

export type PageNode = {
    "@id": string,
    "@type": string,
    url: string,
    mainEntity: any,
    hasPart: any[],
    [key: string]: any
}

export type OnBundleChanged = (isError: boolean, entry: any) => void;

export interface IHtmlSerializer {
    toHtml(renderResult: unknown);
}

export const CACHE_FOLDER = ".sambal";
export const OUTPUT_FOLDER = "public";
export const SAMBAL_ENTRY_FILE = "sambal.entry.js";
export const SAMBAL_SITE_FILE = "sambal.site.js";
export const WEBSOCKET_ADDR = "ws://localhost:3001/";

// collection
export const NULL_VALUE = "__null__";
export const EMPTY_VALUE = "__empty__";
