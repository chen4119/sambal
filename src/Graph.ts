import {
    JSONLD_CONTEXT,
    JSONLD_GRAPH,
    JSONLD_ID,
    JSONLD_TYPE,
    SCHEMA_CONTEXT,
    toJsonLdGraph
} from "sambal-jsonld";
import { isObjectLiteral, writeText } from "./helpers/util";
import { OUTPUT_FOLDER } from "./helpers/constant";
import Media from "./Media";
import Links from "./Links";
import { 
    isExternalSource,
    isImageFile,
    loadRemoteFile,
    loadLocalFile,
    getLocalFilePath,
    isSupportedFile,
    normalizeJsonLdId
} from "./helpers/data"; 


const IMAGE_OBJECT = "imageobject";
const MAX_DEPTH = 100;

export default class Graph {
    private objectCache: Map<string, unknown> = new Map<string, unknown>();
    private links = new Links();
    private blankNodeIndex: number = 1;

    constructor(private baseUrl: string, private media: Media) {

    }

    async serialize() {
        for (const iri of Array.from(this.objectCache.keys())) {
            if (!isExternalSource(iri) && !iri.startsWith("_:")) {
                await writeText(`./${OUTPUT_FOLDER}/content/${iri}.json`, JSON.stringify({
                    [JSONLD_CONTEXT]: {
                        "@vocab": SCHEMA_CONTEXT,
                        "@base": this.baseUrl
                    },
                    ...this.objectCache.get(iri) as object
                }, null, 4));
            }
        }
    }

    async load(src: any) {
        let jsonld = src;
        if (typeof(src) === "string") {
            jsonld = await this.loadJsonLdPath(src);
        }
        await this.ensureJsonLd(jsonld);
        
        jsonld = await this.hydrate(null, null, jsonld, 1);
        this.cacheJsonLd(jsonld);

        if (!jsonld[JSONLD_CONTEXT]) {
            return {
                [JSONLD_CONTEXT]: SCHEMA_CONTEXT,
                ...jsonld,
            };
        }
        return jsonld;
    }

    private cacheJsonLd(jsonld: object) {
        // TODO: What if jsonld already has @graph
        const context = jsonld[JSONLD_CONTEXT] ? jsonld[JSONLD_CONTEXT] : SCHEMA_CONTEXT;
        const flatten = toJsonLdGraph([jsonld], context);
        for (const item of flatten[JSONLD_GRAPH]) {
            this.objectCache.set(item[JSONLD_ID], item);
        }
    }

    private async loadJsonLdPath(src: string) {
        const idFromSrc = normalizeJsonLdId(src);
        if (this.objectCache.has(idFromSrc)) {
            return this.objectCache.get(idFromSrc);
        }

        let jsonld;
        if (isImageFile(src)) {
            // src can be remote or local
            jsonld = await this.media.loadImagePath(src);
        } else if (isExternalSource(src)) {
            jsonld = await loadRemoteFile(src);
        } else {
            let filePath = src;
            // Most likely no file extension specified
            if (!isSupportedFile(filePath)) {
                filePath = getLocalFilePath(src);
            }
            if (isImageFile(filePath)) {
                jsonld = await this.media.loadImagePath(filePath);
            } else {
                jsonld = await loadLocalFile(filePath);
            }
        }
        await this.ensureJsonLd(jsonld, idFromSrc);
        // this.objectCache.set(jsonld[JSONLD_ID], jsonld);
        return jsonld;
    }
    
    // all data need @id and @type
    // make sure image obj are populated correctly
    private async ensureJsonLd(jsonld: any, impliedId?: string) {
        if (!jsonld[JSONLD_ID]) {
            jsonld[JSONLD_ID] = impliedId ? impliedId : `_:${this.blankNodeIndex++}`;
        }
        if (!jsonld[JSONLD_TYPE]) {
            jsonld[JSONLD_TYPE] = "Thing"; // root type
        }
        if (jsonld[JSONLD_TYPE].toLowerCase() === IMAGE_OBJECT) {
            await this.media.loadImageObject(jsonld);
        }
    }

    private async hydrate(
        subjectIRI: string,
        predicateIRI: string,
        target: unknown,
        level: number,
        graph?: Map<string, unknown>
    ) {
        if (level > MAX_DEPTH) {
            throw new Error("Hydrating jsonld exceeded max depth level");
        }

        const nextLevel = level + 1;
        if (Array.isArray(target)) {
            const resolvedArr = [];
            for (const item of target) {
                resolvedArr.push(
                    await this.hydrate(subjectIRI, predicateIRI, item, nextLevel, graph)
                );
            }
            return resolvedArr;
        } else if (this.isJsonLdRef(target)) {
            this.links.add(subjectIRI, predicateIRI, target[JSONLD_ID]);
            let nextTarget;
            if (graph && graph.has(target[JSONLD_ID])) {
                nextTarget = graph.get(target[JSONLD_ID]);
            } else {
                nextTarget = await this.loadJsonLdPath(target[JSONLD_ID])
            }
            return await this.hydrate(
                subjectIRI,
                predicateIRI,
                nextTarget,
                level,
                graph
            );
        } else if (isObjectLiteral(target) && !this.objectCache.has(target[JSONLD_ID])) {
            let objectGraph = null;
            for (const fieldName of Object.keys(target)) {
                if (fieldName !== JSONLD_ID && fieldName !== JSONLD_TYPE && fieldName !== JSONLD_CONTEXT) {
                    const fieldValue = target[fieldName];
                    if (fieldName === JSONLD_GRAPH) {
                        objectGraph = this.getGraphContext(fieldValue);
                    }
                    target[fieldName] = await this.hydrate(
                        target[JSONLD_ID],
                        `schema:${fieldName}`, // TODO: may not always be schema
                        fieldValue,
                        nextLevel,
                        objectGraph ? objectGraph : graph
                    );
                }
            }
        }
        return target;
    }

    private getGraphContext(graph: any[]) {
        const context = new Map<string, unknown>();
        for (const item of graph) {
            context.set(item[JSONLD_ID], item);
        }
        return context;
    }

    private isJsonLdRef(value) {
        return isObjectLiteral(value) && Object.keys(value).length === 1 && typeof (value[JSONLD_ID]) === "string";
    }
}