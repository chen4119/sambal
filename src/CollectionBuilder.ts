import { expandJsonLd, JSONLD_ID, JSONLD_TYPE } from "sambal-jsonld";
import Graph from "./Graph";
import {
    Collection,
    SortBy,
    PartitionKey,
    SORT_ASC
} from "./helpers/constant";
import {
    isJsDate,
    isNullOrUndefined
} from "./helpers/util";
import {
    getSortKey
} from "./helpers/collection";
import { searchLocalFiles, normalizeJsonLdId } from "./helpers/data";
import { isObjectLiteral } from "sambal-jsonld/dist/utils";

type IndexItem = {
    [JSONLD_ID]: string,
    sortValue?: string | number | Date
};

type IndexList = IndexItem[];

type Partition = {
    groupBy?: PartitionKey,
    feed: IndexList
};

const RESERVED_PARTITION_KEY_NAMES = ["pageNum"];

export default class CollectionBuilder {
    private collectionMap: Map<String, Partition[]>;
    constructor(private siteGraph: Graph, private collections: Collection[]) {
        if (collections) {
            collections.forEach(c => {
                c[JSONLD_ID] = normalizeJsonLdId(c[JSONLD_ID]);
            });
        }
        this.collectionMap = new Map<String, Partition[]>();
    }

    async getCollectionPages(collectionIRI: string, pageSize: number) {
        const normalizeIRI = normalizeJsonLdId(collectionIRI);
        const collection = this.getCollection(normalizeIRI);
        let partitions: Partition[] = this.collectionMap.get(normalizeIRI);
        if (!partitions) {
            partitions = await this.build(collection);
        }
        let pages = [];
        for (const partition of partitions) {
            pages.push({
                key: partition.groupBy ? partition.groupBy : null,
                pages: await this.paginatePartition(collection, partition, pageSize)
            });
        }
        return pages;
    }

    async getPartitionPages(collectionIRI: string, partitionKey: PartitionKey, pageSize: number) {
        const normalizeIRI = normalizeJsonLdId(collectionIRI);
        const collection = this.getCollection(normalizeIRI);
        let partitions: Partition[] = this.collectionMap.get(normalizeIRI);
        if (!partitions) {
            partitions = await this.build(collection);
        }
        const keyStrToMatch = this.stringifyKey(partitionKey);
        let pages = [];
        for (const partition of partitions) {
            if (partition.groupBy && this.stringifyKey(partition.groupBy) === keyStrToMatch) {
                pages.push({
                    key: partition.groupBy ? partition.groupBy : null,
                    pages: await this.paginatePartition(collection, partition, pageSize)
                });
            }
        }
        return pages;
    }

    private async paginatePartition(collection: Collection, partition: Partition, pageSize: number) {
        const pages = [];
        if (pageSize > partition.feed.length) {
            const pageIRI = this.getCollectionPageUrl(collection[JSONLD_ID], partition.groupBy);
            pages.push(this.serializeItemList(collection, pageIRI, partition.feed));
        } else {
            let currentIndex = 0;
            let prevItem = null;
            while(currentIndex < partition.feed.length) {
                const start = currentIndex;
                const end = Math.min(start + pageSize, partition.feed.length);
                const position = pages.length;
                const slice = partition.feed.slice(start, end);
                const item = this.serializeListItem(collection, partition, slice, prevItem, position);
                prevItem = item;
                pages.push(item);
                currentIndex = end;
            }
        }
        for (const page of pages) {
            await this.siteGraph.load(page);
        }
        return pages;
    }

    private serializeListItem(collection: Collection, partition: Partition, slice: IndexItem[], prevItem: any, position: number) {
        const pageIRI = this.getCollectionPageUrl(collection[JSONLD_ID], partition.groupBy, position);
        const item: any = {
            [JSONLD_ID]: pageIRI,
            [JSONLD_TYPE]: "ListItem",
            item: this.serializeItemList(collection, `${pageIRI}/items`, slice),
            position: position
        };
        if (prevItem) {
            item.previousItem = {
                [JSONLD_ID]: prevItem[JSONLD_ID]
            };
            prevItem.nextItem = {
                [JSONLD_ID]: item[JSONLD_ID]
            };
        }
        return item;
    }

    private serializeItemList(collection: Collection, listIRI: string, list: IndexItem[]) {
        const listType = collection[JSONLD_TYPE] ? collection[JSONLD_TYPE] : "ItemList";
        return {
            [JSONLD_ID]: listIRI,
            [JSONLD_TYPE]: listType,
            itemListElement: list.map(d => ({
                [JSONLD_ID]: d[JSONLD_ID]
            }))
        }
    }

    private getCollectionPageUrl(collectionIRI: string, partitionKey?: PartitionKey, pageIndex?: number) {
        const paths = [collectionIRI];
        const params: any = partitionKey ? partitionKey : {};
        if (pageIndex) {
            params.pageNum = pageIndex + 1;
        }
        let queries = [];
        for (const name of Object.keys(params)) {
            queries.push(`${encodeURIComponent(name)}=${encodeURIComponent(params[name])}`);
        }
        if (queries.length > 0) {
            return `${collectionIRI}?${queries.join("&")}`;
        }
        return collectionIRI;
    }

    private async build(collection: Collection) {
        const matches = searchLocalFiles(collection.src);
        let partitions: Partition[];
        if (collection.groupBy) {
            partitions = await this.addToPartitionedList(collection, matches);
        } else {
            partitions = await this.addToSingleList(collection, matches);
        }

        if (collection.sortBy) {
            this.sortFeeds(partitions, collection.sortBy);
        }
        this.collectionMap.set(collection[JSONLD_ID], partitions);
        return partitions;
    }

    private async addToSingleList(collection: Collection, filePaths: string[]) {
        const feed: IndexList = [];
        for (const filePath of filePaths) {
            const jsonld = expandJsonLd(await this.siteGraph.load(filePath));
            feed.push(this.getFeedItem(jsonld, collection.sortBy));
        }
        return [{
            feed: feed
        }];
    }

    private async addToPartitionedList(collection: Collection, filePaths: string[]) {
        const partitionMap = new Map<string, Partition>();
        for (const filePath of filePaths) {
            const mainEntity = await this.siteGraph.load(filePath);
            const partitions = collection.groupBy(mainEntity);
            if (Array.isArray(partitions)) {
                for (const partitionKey of partitions) {
                    const partition = this.getPartition(partitionMap, partitionKey);
                    partition.feed.push(this.getFeedItem(mainEntity, collection.sortBy));
                }
            } else if (isObjectLiteral(partitions)) {
                const partition = this.getPartition(partitionMap, partitions);
                partition.feed.push(this.getFeedItem(mainEntity, collection.sortBy));
            } else {
                throw new Error(`Invalid partition key: ${partitions}`);
            }
        }
        return Array.from(partitionMap.values());
    }

    private getPartition(partitionMap: Map<string, Partition>, key: PartitionKey) {
        const keyStr = this.stringifyKey(key);
        let partition: Partition;
        if (partitionMap.has(keyStr)) {
            partition = partitionMap.get(keyStr);
        } else {
            partition = {
                groupBy: key,
                feed: []
            };
            partitionMap.set(keyStr, partition);
        }
        return partition;
    }

    private getFeedItem(jsonld: any, sortBy?: SortBy) {
        const item: IndexItem = { [JSONLD_ID]: jsonld[JSONLD_ID] };
        if (sortBy) {
            item.sortValue = getSortKey(jsonld, sortBy.prop);
        }
        return item;
    }
 
    private sortFeeds(partitions: Partition[], sortBy: SortBy) {
        for (const partition of partitions) {
            partition.feed.sort((a, b) => this.compare(a.sortValue, b.sortValue, sortBy.order));
        }
    }

    private compare(aValue: any, bValue: any, order: string) {
        if (isNullOrUndefined(aValue) && isNullOrUndefined(bValue)) {
            return 0;
        } else if (isNullOrUndefined(aValue)) {
            return 1;
        } else if (isNullOrUndefined(bValue)) {
            return -1;
        }
        return this.compareNonNullValues(aValue, bValue, order);
    }

    private compareNonNullValues(aValue: any, bValue: any, order: string) {
        if (isJsDate(aValue) && isJsDate(bValue)) {
            return order === SORT_ASC ? (aValue.getTime() - bValue.getTime()) : (bValue.getTime() - aValue.getTime());
        } else if (typeof(aValue) === "number" && typeof(bValue) === "number") {
            return order === SORT_ASC ? (aValue - bValue) : (bValue - aValue);
        }
        return order === SORT_ASC ? String(aValue).localeCompare(String(bValue)) : String(bValue).localeCompare(String(aValue));
    }

    private getCollection(collectionIRI: string) {
        const collection = this.collections ? this.collections.find(d => d[JSONLD_ID] === collectionIRI) : null;
        if (!collection) {
            throw new Error(`collection ${collectionIRI} not found`);
        }
        return collection;
    }

    private stringifyKey(partitionKey: PartitionKey) {
        const fieldNames = Object.keys(partitionKey);
        fieldNames.sort();
        let keyStr = "";
        for (const name of fieldNames) {
            if (RESERVED_PARTITION_KEY_NAMES.indexOf(name.toLowerCase()) >= 0) {
                throw new Error(`${name} is a reserved partition key name`);
            }
            if (typeof(partitionKey[name]) !== "string" && 
                typeof(partitionKey[name]) !== "number" &&
                !isJsDate(partitionKey[name])) {
                throw new Error(`Invalid partition key: ${JSON.stringify(partitionKey)} - Only string, number or date value allowed`);
            }
            keyStr += `${name}=${partitionKey[name]};`;
        }
        return keyStr;
    }

}