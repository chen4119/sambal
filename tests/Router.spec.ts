import Router from "../src/Router";
import CollectionBuilder from "../src/CollectionBuilder";
import Graph from "../src/Graph";
import Media from "../src/Media";
import Links from "../src/Links";
import { Collection } from "../src/helpers/constant";

describe("SambalCollection", () => {
    const baseUrl = "https://example.com";
    let router: Router;
    let graph: Graph;
    let links: Links;
    let collectionBuilder: CollectionBuilder;

    const collections: Collection[] = [
        {
            "@id": "collections/year",
            "@type": "SiteNavigationElement",
            src: ["blogs/**/*"],
            sortBy: {
                prop: "dateModified",
                order: "asc"
            }
        }
    ];

    beforeEach(async () => {
        links = new Links();
        collectionBuilder = new CollectionBuilder(collections);
        graph = new Graph(baseUrl, new Media([]), links, collectionBuilder);
        router = new Router(graph, collectionBuilder);

        router.instance
        .landingPage("johnsmith", {page: {
            hasPart: {"@id": "collections/year"}
        }})
        .itemPage("/blog1", "blogs/blog1")
        .itemPage("/canonical/blog1", "blogs/blog1", { canonical: true })
        .itemPage("/blog2", "blogs/blog2");
    });

    it('getRoutes', async () => {
        const pages = await router.getRoutes();
        expect(pages.length).toBe(4);
    });

    it('landing page has sitenav', async () => {
        const pages = await router.getRoutes();
        const landing = pages.find(p => p.url === "/");
        expect(landing.hasPart.length).toBe(2);
    });

    it('set canonical url for blogs/blog1', async () => {
        const pages = await router.getRoutes();
        const blog1 = pages.filter(p => p.url === "/blog1")[0];
        expect(blog1.mainEntity.mainEntityOfPage).toBe("/canonical/blog1");

        const canonical = pages.filter(p => p.url === "/canonical/blog1")[0];
        expect(canonical.mainEntity.mainEntityOfPage).toBe("/canonical/blog1");
    });


});