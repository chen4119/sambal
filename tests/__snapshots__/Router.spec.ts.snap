// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router get / 1`] = `
Object {
  "@type": "WebPage",
  "hasPart": Object {
    "@id": "/header",
    "@type": "WPHeader",
    "hasPart": Array [
      Object {
        "@type": "SiteNavigationElement",
        "name": "About",
        "url": "/about",
      },
      Object {
        "@type": "SiteNavigationElement",
        "name": "Docs",
        "url": "/get-started",
      },
      Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "My first blog",
          "url": "/blogs/blog1",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "My second blog",
          "url": "/blogs/blog2",
        },
      ],
    ],
    "headline": "Sambal",
    "potentialAction": Array [
      Object {
        "@type": "DownloadAction",
        "target": "https://github.com/chen4119/sambal",
      },
    ],
  },
  "mainEntity": Object {
    "@id": "/",
    "@type": "SoftwareSourceCode",
    "description": "A linked data static site generator",
    "headline": "A linked data static site generator",
    "name": "Sambal",
    "text": "<need to <> escape>",
  },
  "mainEntityOfPage": "/",
  "url": "/",
}
`;

exports[`Router get /alias/subfolder/blog1 1`] = `
Object {
  "@type": "WebPage",
  "description": "alias folder",
  "mainEntity": Object {
    "@id": "/blogs/blog1",
    "@type": "BlogPosting",
    "author": Object {
      "@id": "/john%20smith",
      "@type": "Person",
      "email": "smitty123@hotmail.com",
      "familyName": "Smith",
      "givenName": "John",
      "name": "John Smith",
    },
    "dateCreated": 2019-10-26T00:00:00.000Z,
    "dateModified": 2019-10-26T00:00:00.000Z,
    "encodingFormat": "text/markdown",
    "headline": "My first blog",
    "keywords": Array [
      "rxjs",
      "java script",
    ],
    "position": 1,
    "text": "

Hello world",
  },
  "mainEntityOfPage": "/blogs/blog1",
  "url": "/alias/subfolder/blog1",
}
`;

exports[`Router get /archive/2017/1 1`] = `
Object {
  "@type": "WebPage",
  "hasPart": Object {
    "@id": "/header",
    "@type": "WPHeader",
    "hasPart": Array [
      Object {
        "@type": "SiteNavigationElement",
        "name": "About",
        "url": "/about",
      },
      Object {
        "@type": "SiteNavigationElement",
        "name": "Docs",
        "url": "/get-started",
      },
      Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "My first blog",
          "url": "/blogs/blog1",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "My second blog",
          "url": "/blogs/blog2",
        },
      ],
    ],
    "headline": "Sambal",
    "potentialAction": Array [
      Object {
        "@type": "DownloadAction",
        "target": "https://github.com/chen4119/sambal",
      },
    ],
  },
  "mainEntity": Object {
    "@type": "ListItem",
    "item": Object {
      "@id": "/collections/year/_part/year=2017/1",
      "@type": "ItemList",
      "itemListElement": Array [
        Object {
          "@id": "/blogs/blog2",
          "@type": "BlogPosting",
          "author": Object {
            "@id": "/john%20smith",
            "@type": "Person",
            "email": "smitty123@hotmail.com",
            "familyName": "Smith",
            "givenName": "John",
            "name": "John Smith",
          },
          "dateCreated": 2017-10-26T00:00:00.000Z,
          "dateModified": 2017-10-26T00:00:00.000Z,
          "encodingFormat": "text/markdown",
          "headline": "My second blog",
          "keywords": Array [
            "sambal",
            "java script",
          ],
          "position": 2,
          "text": "

Hello world2",
        },
      ],
    },
    "position": 0,
  },
  "mainEntityOfPage": "/archive/2017/1",
  "url": "/archive/2017/1",
}
`;

exports[`Router get /archive/org-a 1`] = `
Object {
  "@type": "WebPage",
  "hasPart": Object {
    "@id": "/header",
    "@type": "WPHeader",
    "hasPart": Array [
      Object {
        "@type": "SiteNavigationElement",
        "name": "About",
        "url": "/about",
      },
      Object {
        "@type": "SiteNavigationElement",
        "name": "Docs",
        "url": "/get-started",
      },
      Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "My first blog",
          "url": "/blogs/blog1",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "My second blog",
          "url": "/blogs/blog2",
        },
      ],
    ],
    "headline": "Sambal",
    "potentialAction": Array [
      Object {
        "@type": "DownloadAction",
        "target": "https://github.com/chen4119/sambal",
      },
    ],
  },
  "mainEntity": Object {
    "@id": "/archive/org-a",
    "@type": "Organization",
    "address": Object {
      "@type": "PostalAddress",
      "postalCode": 12345,
      "streetAddress": "1 main st",
    },
    "identifier": "org-a",
    "name": "org a",
    "telephone": "123-435-2334",
  },
  "mainEntityOfPage": "/archive/org-a",
  "url": "/archive/org-a",
}
`;

exports[`Router get /blogs/blog1 1`] = `
Object {
  "@type": "WebPage",
  "hasPart": Array [
    Object {
      "@type": "SiteNavigationElement",
      "name": "About",
      "url": "/about",
    },
  ],
  "mainEntity": Object {
    "@id": "/blogs/blog1",
    "@type": "BlogPosting",
    "author": Object {
      "@id": "/john%20smith",
      "@type": "Person",
      "email": "smitty123@hotmail.com",
      "familyName": "Smith",
      "givenName": "John",
      "name": "John Smith",
    },
    "dateCreated": 2019-10-26T00:00:00.000Z,
    "dateModified": 2019-10-26T00:00:00.000Z,
    "encodingFormat": "text/markdown",
    "headline": "My first blog",
    "keywords": Array [
      "rxjs",
      "java script",
    ],
    "position": 1,
    "text": "

Hello world",
  },
  "mainEntityOfPage": "/blogs/blog1",
  "url": "/blogs/blog1",
}
`;

exports[`Router iterate routes 1`] = `
Array [
  Object {
    "@type": "WebPage",
    "hasPart": Object {
      "@id": "/header",
      "@type": "WPHeader",
      "hasPart": Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "About",
          "url": "/about",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "Docs",
          "url": "/get-started",
        },
        Array [
          Object {
            "@type": "SiteNavigationElement",
            "name": "My first blog",
            "url": "/blogs/blog1",
          },
          Object {
            "@type": "SiteNavigationElement",
            "name": "My second blog",
            "url": "/blogs/blog2",
          },
        ],
      ],
      "headline": "Sambal",
      "potentialAction": Array [
        Object {
          "@type": "DownloadAction",
          "target": "https://github.com/chen4119/sambal",
        },
      ],
    },
    "mainEntity": Object {
      "@id": "/",
      "@type": "SoftwareSourceCode",
      "description": "A linked data static site generator",
      "headline": "A linked data static site generator",
      "name": "Sambal",
      "text": "<need to <> escape>",
    },
    "mainEntityOfPage": "/",
    "url": "/",
  },
  Object {
    "@type": "WebPage",
    "hasPart": Array [
      Object {
        "@type": "SiteNavigationElement",
        "name": "About",
        "url": "/about",
      },
    ],
    "mainEntity": Object {
      "@id": "/blogs/blog1",
      "@type": "BlogPosting",
      "author": Object {
        "@id": "/john%20smith",
        "@type": "Person",
        "email": "smitty123@hotmail.com",
        "familyName": "Smith",
        "givenName": "John",
        "name": "John Smith",
      },
      "dateCreated": 2019-10-26T00:00:00.000Z,
      "dateModified": 2019-10-26T00:00:00.000Z,
      "encodingFormat": "text/markdown",
      "headline": "My first blog",
      "keywords": Array [
        "rxjs",
        "java script",
      ],
      "position": 1,
      "text": "

Hello world",
    },
    "mainEntityOfPage": "/blogs/blog1",
    "url": "/blogs/blog1",
  },
  Object {
    "@type": "WebPage",
    "hasPart": Array [
      Object {
        "@type": "SiteNavigationElement",
        "name": "About",
        "url": "/about",
      },
    ],
    "mainEntity": Object {
      "@id": "/blogs/blog2",
      "@type": "BlogPosting",
      "author": Object {
        "@id": "/john%20smith",
        "@type": "Person",
        "email": "smitty123@hotmail.com",
        "familyName": "Smith",
        "givenName": "John",
        "name": "John Smith",
      },
      "dateCreated": 2017-10-26T00:00:00.000Z,
      "dateModified": 2017-10-26T00:00:00.000Z,
      "encodingFormat": "text/markdown",
      "headline": "My second blog",
      "keywords": Array [
        "sambal",
        "java script",
      ],
      "position": 2,
      "text": "

Hello world2",
    },
    "mainEntityOfPage": "/blogs/blog2",
    "url": "/blogs/blog2",
  },
  Object {
    "@type": "WebPage",
    "hasPart": Object {
      "@id": "/header",
      "@type": "WPHeader",
      "hasPart": Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "About",
          "url": "/about",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "Docs",
          "url": "/get-started",
        },
        Array [
          Object {
            "@type": "SiteNavigationElement",
            "name": "My first blog",
            "url": "/blogs/blog1",
          },
          Object {
            "@type": "SiteNavigationElement",
            "name": "My second blog",
            "url": "/blogs/blog2",
          },
        ],
      ],
      "headline": "Sambal",
      "potentialAction": Array [
        Object {
          "@type": "DownloadAction",
          "target": "https://github.com/chen4119/sambal",
        },
      ],
    },
    "mainEntity": Object {
      "@id": "/archive/org-a",
      "@type": "Organization",
      "address": Object {
        "@type": "PostalAddress",
        "postalCode": 12345,
        "streetAddress": "1 main st",
      },
      "identifier": "org-a",
      "name": "org a",
      "telephone": "123-435-2334",
    },
    "mainEntityOfPage": "/archive/org-a",
    "url": "/archive/org-a",
  },
  Object {
    "@type": "WebPage",
    "hasPart": Object {
      "@id": "/header",
      "@type": "WPHeader",
      "hasPart": Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "About",
          "url": "/about",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "Docs",
          "url": "/get-started",
        },
        Array [
          Object {
            "@type": "SiteNavigationElement",
            "name": "My first blog",
            "url": "/blogs/blog1",
          },
          Object {
            "@type": "SiteNavigationElement",
            "name": "My second blog",
            "url": "/blogs/blog2",
          },
        ],
      ],
      "headline": "Sambal",
      "potentialAction": Array [
        Object {
          "@type": "DownloadAction",
          "target": "https://github.com/chen4119/sambal",
        },
      ],
    },
    "mainEntity": Object {
      "@id": "/archive/org-b",
      "@type": "Organization",
      "identifier": "org-b",
      "name": "org b",
      "telephone": "123-435-1564",
    },
    "mainEntityOfPage": "/archive/org-b",
    "url": "/archive/org-b",
  },
  Object {
    "@type": "WebPage",
    "hasPart": Object {
      "@id": "/header",
      "@type": "WPHeader",
      "hasPart": Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "About",
          "url": "/about",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "Docs",
          "url": "/get-started",
        },
        Array [
          Object {
            "@type": "SiteNavigationElement",
            "name": "My first blog",
            "url": "/blogs/blog1",
          },
          Object {
            "@type": "SiteNavigationElement",
            "name": "My second blog",
            "url": "/blogs/blog2",
          },
        ],
      ],
      "headline": "Sambal",
      "potentialAction": Array [
        Object {
          "@type": "DownloadAction",
          "target": "https://github.com/chen4119/sambal",
        },
      ],
    },
    "mainEntity": Object {
      "@type": "ListItem",
      "item": Object {
        "@id": "/collections/year/_part/year=2019/1",
        "@type": "ItemList",
        "itemListElement": Array [
          Object {
            "@id": "/blogs/blog1",
            "@type": "BlogPosting",
            "author": Object {
              "@id": "/john%20smith",
              "@type": "Person",
              "email": "smitty123@hotmail.com",
              "familyName": "Smith",
              "givenName": "John",
              "name": "John Smith",
            },
            "dateCreated": 2019-10-26T00:00:00.000Z,
            "dateModified": 2019-10-26T00:00:00.000Z,
            "encodingFormat": "text/markdown",
            "headline": "My first blog",
            "keywords": Array [
              "rxjs",
              "java script",
            ],
            "position": 1,
            "text": "

Hello world",
          },
        ],
      },
      "position": 0,
    },
    "mainEntityOfPage": "/archive/2019/1",
    "url": "/archive/2019/1",
  },
  Object {
    "@type": "WebPage",
    "hasPart": Object {
      "@id": "/header",
      "@type": "WPHeader",
      "hasPart": Array [
        Object {
          "@type": "SiteNavigationElement",
          "name": "About",
          "url": "/about",
        },
        Object {
          "@type": "SiteNavigationElement",
          "name": "Docs",
          "url": "/get-started",
        },
        Array [
          Object {
            "@type": "SiteNavigationElement",
            "name": "My first blog",
            "url": "/blogs/blog1",
          },
          Object {
            "@type": "SiteNavigationElement",
            "name": "My second blog",
            "url": "/blogs/blog2",
          },
        ],
      ],
      "headline": "Sambal",
      "potentialAction": Array [
        Object {
          "@type": "DownloadAction",
          "target": "https://github.com/chen4119/sambal",
        },
      ],
    },
    "mainEntity": Object {
      "@type": "ListItem",
      "item": Object {
        "@id": "/collections/year/_part/year=2017/1",
        "@type": "ItemList",
        "itemListElement": Array [
          Object {
            "@id": "/blogs/blog2",
            "@type": "BlogPosting",
            "author": Object {
              "@id": "/john%20smith",
              "@type": "Person",
              "email": "smitty123@hotmail.com",
              "familyName": "Smith",
              "givenName": "John",
              "name": "John Smith",
            },
            "dateCreated": 2017-10-26T00:00:00.000Z,
            "dateModified": 2017-10-26T00:00:00.000Z,
            "encodingFormat": "text/markdown",
            "headline": "My second blog",
            "keywords": Array [
              "sambal",
              "java script",
            ],
            "position": 2,
            "text": "

Hello world2",
          },
        ],
      },
      "position": 0,
    },
    "mainEntityOfPage": "/archive/2017/1",
    "url": "/archive/2017/1",
  },
  Object {
    "@type": "WebPage",
    "description": "alias folder",
    "mainEntity": Object {
      "@id": "/blogs/blog1",
      "@type": "BlogPosting",
      "author": Object {
        "@id": "/john%20smith",
        "@type": "Person",
        "email": "smitty123@hotmail.com",
        "familyName": "Smith",
        "givenName": "John",
        "name": "John Smith",
      },
      "dateCreated": 2019-10-26T00:00:00.000Z,
      "dateModified": 2019-10-26T00:00:00.000Z,
      "encodingFormat": "text/markdown",
      "headline": "My first blog",
      "keywords": Array [
        "rxjs",
        "java script",
      ],
      "position": 1,
      "text": "

Hello world",
    },
    "mainEntityOfPage": "/blogs/blog1",
    "url": "/alias/subfolder/blog1",
  },
]
`;
